worker/src/RTC/WebRtcServer.cpp

```diff
@@ -0,0 +1,517 @@
+
+		idx = 0;
+		for (auto& kv : this->mapTupleWebRtcTransport)
+		{
+			const auto& tupleHash       = kv.first;
+			const auto* webRtcTransport = kv.second;
+
+			jsonTupleHashesIt->emplace_back(json::value_t::object);
+
+			auto& jsonEntry = (*jsonTupleHashesIt)[idx];
+
+			jsonEntry["tupleHash"]         = tupleHash;
+			jsonEntry["webRtcTransportId"] = webRtcTransport->id;
+
+			++idx;
+		}
+	}
+
+	void WebRtcServer::HandleRequest(Channel::ChannelRequest* request)
+	{
+		MS_TRACE();
+
+		switch (request->methodId)
+		{
+			case Channel::ChannelRequest::MethodId::WEBRTC_SERVER_DUMP:
+			{
+				json data = json::object();
+
+				FillJson(data);
+
+				request->Accept(data);
+
+				break;
+			}
+
+			default:
+			{
+				MS_THROW_ERROR("unknown method '%s'", request->method.c_str());
+			}
+		}
+	}
+
+	std::vector<RTC::IceCandidate> WebRtcServer::GetIceCandidates(
+	  bool enableUdp, bool enableTcp, bool preferUdp, bool preferTcp)
+	{
+		MS_TRACE();
+
+		std::vector<RTC::IceCandidate> iceCandidates;
+		uint16_t iceLocalPreferenceDecrement{ 0 };
+
+		for (auto& item : this->udpSocketOrTcpServers)
+		{
+			if (item.udpSocket && enableUdp)
+			{
+				uint16_t iceLocalPreference = IceCandidateDefaultLocalPriority - iceLocalPreferenceDecrement;
+
+				if (preferUdp)
+					iceLocalPreference += 1000;
+
+				uint32_t icePriority = generateIceCandidatePriority(iceLocalPreference);
+
+				if (item.announcedIp.empty())
+					iceCandidates.emplace_back(item.udpSocket, icePriority);
+				else
+					iceCandidates.emplace_back(item.udpSocket, icePriority, item.announcedIp);
+			}
+			else if (item.tcpServer && enableTcp)
+			{
+				uint16_t iceLocalPreference = IceCandidateDefaultLocalPriority - iceLocalPreferenceDecrement;
+
+				if (preferTcp)
+					iceLocalPreference += 1000;
+
+				uint32_t icePriority = generateIceCandidatePriority(iceLocalPreference);
+
+				if (item.announcedIp.empty())
+					iceCandidates.emplace_back(item.tcpServer, icePriority);
+				else
+					iceCandidates.emplace_back(item.tcpServer, icePriority, item.announcedIp);
+			}
+
+			// Decrement initial ICE local preference for next IP.
+			iceLocalPreferenceDecrement += 100;
+		}
+
+		return iceCandidates;
+	}
+
+	inline std::string WebRtcServer::GetLocalIceUsernameFragmentFromReceivedStunPacket(
+	  RTC::StunPacket* packet) const
+	{
+		MS_TRACE();
+
+		// Here we inspect the USERNAME attribute of a received STUN request and
+		// extract its remote usernameFragment (the one given to our IceServer as
+		// local usernameFragment) which is the first value in the attribute value
+		// before the ":" symbol.
+
+		auto& username  = packet->GetUsername();
+		size_t colonPos = username.find(":");
+
+		// If no colon is found just return the whole USERNAME attribute anyway.
+		if (colonPos == std::string::npos)
+			return username;
+
+		return username.substr(0, colonPos);
+	}
+
+	inline void WebRtcServer::OnPacketReceived(RTC::TransportTuple* tuple, const uint8_t* data, size_t len)
```

Comment:
a question about this one
You search for a ufrag in mapLocalIceUsernameFragmentWebRtcTransport first. Isn't it more efficient to search mapTupleWebRtcTransport for the tuple hash first and only fall back to stun (mainly for new connections) when that lookup fails?

Reply:
Actually computing the tiple hash (see TransportTuple.hpp SetHash() method) is less efficient than doing a lookup of a string in a map.

Reply:
Did you look into something specialized like libre's [sa_hash](https://github.com/baresip/re/blob/62f755ce05bd2597f75009a2a6ff4baa2719e568/src/sa/sa.c#L507)? Which might also allow using ordered map.

Reply:
@jmillan 👆🏻👆🏻👆🏻👆🏻👆🏻👆🏻🌽

Reply:
yes, that's what we're looking for

Reply:
@jmillan take into account that such `sa_hash` function just reads a `struct sa` while we have to read two (local and remote). So we should use `uint64_t` instead of `uint32_t` and "concatenate" the resulting numbers:

```
// Pseudo code:
uint64_t hash = (localSaHash << 32) + remoteSaHash);
```

Reply:
Working on https://github.com/versatica/mediasoup/pull/839